File: service_frontend/types/permission.ts
/**
 * Permission catalog types - the dynamic RBAC catalog (plan 03 §2). A permission
 * is a flat key `"<resource>.<action>"`. Modules (core + App-Store) declare their
 * own permissions via a CSV synced into the backend `permissions` table; this is
 * the shape the catalog endpoint returns to drive the role Permissions tab.
 */

/** One grantable action on a resource (e.g. `users.create`). */
export interface PermissionAction {
  /** Flat permission key - `"<resource>.<action>"`. */
  key: string;
  /** Action segment - `read | create | update | delete | <custom>`. */
  action: string;
  /** Human label shown as the dropdown option (e.g. "Create"). */
  actionLabel: string;
  description: string | null;
}

/** A resource (a matrix row) and its available actions, owned by a module. */
export interface PermissionResource {
  /** Resource segment of the key (e.g. `users`). */
  resource: string;
  /** Heading shown for this resource row (e.g. "User Management"). */
  resourceLabel: string;
  /** Owning module - `core` or an App-Store module name (groups the sections). */
  module: string;
  actions: PermissionAction[];
}

/** The catalog, flat list of resources; the UI groups by `module` for display. */
export type PermissionCatalog = PermissionResource[];

File: service_backend/app/schemas/permission.py
"""Permission catalog schemas - the grouped shape the role Permissions tab reads."""
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class PermissionActionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    action: str
    actionLabel: str = Field(validation_alias="action_label")
    description: Optional[str] = None


class PermissionResourceOut(BaseModel):
    """One resource (matrix row) + its available actions, owned by a module."""

    resource: str
    resourceLabel: str
    module: str
    actions: List[PermissionActionOut] = []
